Feature implement missing program rule functions

idelcano opened the pull request:
Closes #255

idelcano committed:
working in new tests

idelcano committed:
Merge branch 'feature-rename_tests' into feature-implement_missing_program_rule_functions

idelcano committed:
fix and add more test in old functions tests

idelcano committed:
Added Zing, Zpvc, Oizp, WeeksBetween and MonthsBetween funcitons

idelcano committed:
Added program rule function condition

idelcano committed:
Added yearbetween function

idelcano committed:
Added modulus

idelcano committed:
Added concatenate function

idelcano committed:
Added Add days function

idelcano committed:
Added count function and fix rulefunction switch

idelcano committed:
add count text and modify RuleVariableValueMapBuilder permissions

idelcano committed:
Added validate pattern function

idelcano committed:
fix comment

idelcano committed:
Added substring function

idelcano committed:
Merge branch 'feature-rename_tests' into feature-implement_missing_program_rule_functions

idelcano committed:
Added split function

idelcano committed:
Added round function

idelcano committed:
Added Right function

idelcano committed:
Added left function

idelcano committed:
Added length function

idelcano committed:
Added last event function

idelcano committed:
refactor constant

idelcano committed:
Added count if zero pos function

idelcano committed:
fix comment

idelcano committed:
Added count if value function

idelcano committed:
fix typo

idelcano committed:
Add hasValue test and fix function

idelcano committed:
Merge branch 'development' into feature-implement_missing_program_rule_functions

xurxodev commented on core-rules/src/main/java/org/hisp/dhis/rules/RuleVariableValue.java:
```diff
@@ -47,7 +47,7 @@ static RuleVariableValue create(@Nonnull String value,
 
     @Nonnull
-    static RuleVariableValue create(@Nonnull String value,
+    public static RuleVariableValue create(@Nonnull String value,
```
We should not change this method to public because It's [private to package](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html). The goal is not to access from outside from package to this class,  only RuleEngine, RuleEngineContext ... can access it. If we change it to public we are breaking encapsulation.

xurxodev commented on core-rules/src/main/java/org/hisp/dhis/rules/RuleVariableValueMapBuilder.java:
```diff
@@ -28,7 +28,7 @@
 
 @SuppressWarnings("PMD.GodClass")
-final class RuleVariableValueMapBuilder {
+public final class RuleVariableValueMapBuilder {
```
We should not change this method to public because It's [private to package](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html). The goal is not to access from outside from package to this class,  only RuleEngine, RuleEngineContext ... can access it. If we change it to public we are breaking encapsulation.

xurxodev commented on core-rules/src/main/java/org/hisp/dhis/rules/RuleVariableValueMapBuilder.java:
```diff
@@ -96,12 +96,12 @@ static RuleVariableValueMapBuilder target(@Nonnull RuleEnrollment ruleEnrollment
 
     @Nonnull
-    static RuleVariableValueMapBuilder target(@Nonnull RuleEvent ruleEvent) {
+    public static RuleVariableValueMapBuilder target(@Nonnull RuleEvent ruleEvent) {
```
We should not change this method to public because It's [private to package](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html). The goal is not to access from outside from package to this class,  only RuleEngine, RuleEngineContext ... can access it. If we change it to public we are breaking encapsulation.

xurxodev commented on core-rules/src/main/java/org/hisp/dhis/rules/RuleVariableValueMapBuilder.java:
```diff
@@ -96,12 +96,12 @@ static RuleVariableValueMapBuilder target(@Nonnull RuleEnrollment ruleEnrollment
 
     @Nonnull
-    RuleVariableValueMapBuilder ruleVariables(@Nonnull List<RuleVariable> ruleVariables) {
+    public RuleVariableValueMapBuilder ruleVariables(@Nonnull List<RuleVariable> ruleVariables) {
```
We should not change this method to public because It's [private to package](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html). The goal is not to access from outside from package to this class,  only RuleEngine, RuleEngineContext ... can access it. If we change it to public we are breaking encapsulation.

xurxodev commented on core-rules/src/main/java/org/hisp/dhis/rules/RuleVariableValueMapBuilder.java:
```diff
@@ -129,7 +129,7 @@ RuleVariableValueMapBuilder ruleEvents(@Nonnull List<RuleEvent> ruleEvents) {
 
     @Nonnull
-    Map<String, RuleVariableValue> build() {
+    public Map<String, RuleVariableValue> build() {
```
We should not change this method to public because It's [private to package](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html). The goal is not to access from outside from package to this class,  only RuleEngine, RuleEngineContext ... can access it. If we change it to public we are breaking encapsulation.

xurxodev reviewed (changes requested):
@idelcano I have reviewed this PR and there are some things to change:
- You are changing some private to package classes and functions to public and we should not do this because we breaking the encapsulation.
- New implemented functions should be private to package like existed functions.
- I think that integration tests with RuleEngine are necessary because if for example, I change  D2_LEFT to "d2:leftndhdc" all test passed but this is wrong and it will not work.

idelcano committed:
Merge branch 'development' into feature-implement_missing_program_rule_functions

idelcano committed:
fix tests

idelcano committed:
modify rule function last event date

idelcano committed:
fix some lint results

idelcano committed:
Add RuleFunction should evaluate correct rule new functions assertions

xurxodev committed:
Merge branch 'development' into feature-implement_missing_program_rule_functions

xurxodev committed:
refactor zing function to more readable code, fix bugs and add more tests cases

xurxodev committed:
refactor oizp code to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor zpvc func to more readable, fix bugs and add more tests cases

xurxodev committed:
remove condition function, reasons:
 - does not exists in tracker web
 - the structure of this function does not fit with the new SDK

xurxodev committed:
refactor daysBetween func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor weeksBetween func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor monthsBetween func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor yearsBetween func to more readable, fix bugs and add more tests cases

xurxodev committed:
add negative results test cases to date functions

xurxodev committed:
refactor floor func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor modulus func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor concatenate func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor addDays func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor ceil func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor round func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor validatePattern func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor left func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor right func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor length func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor split func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor substring func to more readable, fix bugs and add more tests cases

xurxodev committed:
refactor hasValue func to more readable, fix bugs and add more tests cases

xurxodev committed:
remove LastEventDate function, reasons:
 - does not exists in tracker web
 - does not exists in dhis2 documentation

xurxodev committed:
refactor count func to more readable, fix bugs and add more tests

xurxodev committed:
refactor countIfValue func to more readable, fix bugs and add more tests

xurxodev committed:
refactor countIfZeroPos func to more readable, fix bugs and add more tests

xurxodev reviewed (approved):
thanks @idelcano. I have solved many errors and I have refactored enough code. When you have time to review it and let me know if you have any questions.
